File: exercises-scala/src/test/scala/nl/trivento/propertybased/exercise1/SumTest.scala
package nl.trivento.propertybased.exercise1

import org.scalacheck.Gen
import org.scalatest.prop.GeneratorDrivenPropertyChecks
import org.scalatest.{FlatSpec, Matchers}


class SumTest extends FlatSpec with Matchers with GeneratorDrivenPropertyChecks {


  /**
    * @return The two numbers a and b added together
    */
  def sum(a: Int, b: Int): Int = a + b

  behavior of "sum"

  // Traditional example based test
  it should "correctly add numbers" in {
    sum(1,2) shouldBe 3
    sum(123,0) shouldBe 123
    sum(123,-100) shouldBe 23
  }


  // Property based tests

  it should "execute without crashing" in {
    // This is an example of how to create a test with takes two randomly generated numbers as an argument
    // This test itself does not validate anything, however sometimes we just want to verify that a method terminated and no exceptions are thrown
    forAll { (a: Int, b: Int) =>
      sum(a, b)
    }
  }


  // Example how to generate number in a range
  it should "generate numbers in a range" in {
    val numberGenerator = Gen.chooseNum(1, 10)
    forAll(numberGenerator) { a: Int =>
      a should be >= 1
      a should be <= 10
    }
  }


  it should "be cummutative" in {
    forAll { (a: Int, b: Int) =>
      sum(a, b) shouldBe sum(b, a)
    }
  }

  it should "be associative" in {
    forAll { (a: Int, b: Int, c: Int) =>
      sum(sum(a, b), c) shouldBe sum(a, sum(b, c))
    }
  }

  it should "allow adding 0 (idenity property)" in {
    forAll { a: Int =>
      sum(a, 0) shouldBe a
    }
  }

  it should "be return 0 when adding the inverse" in {
    forAll { a: Int =>
      sum(a, -a) shouldBe 0
    }
  }

  // This last test fails if when very high numbers are used
  it should "return a positive outcome when two positive numbers are added" in {
    val genInt = Gen.chooseNum(0, Int.MaxValue / 2)
    forAll(genInt, genInt) { (a: Int, b: Int) =>
      whenever(a >= 0 && b >= 0) {
        sum(a, b) should be >= 0
      }
    }
  }

}

File: exercises-java/src/test/java/nl/trivento/propertybased/exercise1/SumTest.java
package nl.trivento.propertybased.exercise1;

import com.pholser.junit.quickcheck.Property;
import com.pholser.junit.quickcheck.generator.InRange;
import com.pholser.junit.quickcheck.runner.JUnitQuickcheck;
import org.junit.Test;
import org.junit.runner.RunWith;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

@RunWith(JUnitQuickcheck.class)
public class SumTest {

    /**
     * @return The two numbers a and b added together
     */
    public static int sum(int a, int b) {
        return a + b;
    }

    @Test public void exampleBasedTest() {
        assertEquals(3, sum(1,2));
        assertEquals(123, sum(123,0));
        assertEquals(23, sum(123,-100));
    }

    @Property public void exampleProperty(int a, int b) {
        // This is an example of how to create a test with takes two randomly generated numbers as an argument
        // This test itself does not validate anything, however sometimes we just want to verify that a method terminated and no exceptions are thrown
        sum(a, b);
    }

    @Property public void rangeProperty(@InRange(minInt = 1, maxInt = 10) int a) {
        // Sometimes we need to generate number which are within a specific range, the @InRange annotation can help us with this
        assertTrue("The generated number should be in the specified range", 1 <= a && a <= 10);
    }


    @Property public void cummutativeProperty(int a, int b) {
        assertEquals(sum(a, b), sum(b, a));
    }

    @Property public void associativeProperty(int a, int b, int c) {
        assertEquals(sum(sum(a, b), c), sum(a, sum(b, c)));
    }

    @Property public void identityProperty(int a) {
        assertEquals(sum(a, 0), a);
    }

    @Property public void inverseProperty(int a) {
        assertEquals(sum(a, -a), 0);
    }


    //     This last test fails if no range annotation is used
    @Property public void addingPositiveIntegersShouldBePositive(@InRange(minInt = 0, maxInt = Integer.MAX_VALUE / 2) int a, @InRange(minInt = 0, maxInt = Integer.MAX_VALUE / 2) int b) {
        assertTrue(sum(a, b) >= 0);
    }
}